### ISSUE TITLE
TCP connection close after 60s

### ISSUE THREAD

**Issue by sunboshan** [pma/amqp#88]
RabbitMQ 3.7.3
AMQP 1.0.0
Erlang 20.2.2
Elixir 1.6.0

Start a Rabbit MQ server locally, then use amqp client to connect to it.

```
Erlang/OTP 20 [erts-9.2] [source] [64-bit] [smp:4:4] [ds:4:4:10] [async-threads:10] [hipe] [kernel-poll:false] [dtrace]
Interactive Elixir (1.6.0) - press Ctrl+C to exit (type h() ENTER for help)

iex(1)> AMQP.Connection.open()
{:ok, %AMQP.Connection{pid: #PID<0.223.0>}}
iex(2)> # wait 60s
21:37:31.909 [error] GenServer #PID<0.227.0> terminating
** (stop) {:socket_error, :timeout}
Last message: {:inet_async, #Port<0.6798>, 5, {:error, :timeout}}
State: {:state, #Port<0.6798>, #PID<0.223.0>, #PID<0.225.0>, {:method, :rabbit_framing_amqp_0_9_1}, {:expecting_header, ""}}

21:37:31.925 [error] GenServer #PID<0.223.0> terminating
** (stop) {:socket_error, :timeout}
Last message: {:socket_error, :timeout}
State: {:state, :amqp_network_connection, {:state, #Port<0.6798>, "client 127.0.0.1:50526 -> 127.0.0.1:5672", 60, #PID<0.226.0>, 131072, #PID<0.222.0>, :undefined, false}, #PID<0.225.0>, {:amqp_params_network, "guest", "guest", "/", 'localhost', 5672, 0, 0, 0, :infinity, :none, [&:amqp_auth_mechanisms.plain/3, &:amqp_auth_mechanisms.amqplain/3], [], []}, 0, [{"capabilities", :table, [{"publisher_confirms", :bool, true}, {"exchange_exchange_bindings", :bool, true}, {"basic.nack", :bool, true}, {"consumer_cancel_notify", :bool, true}, {"connection.blocked", :bool, true}, {"consumer_priorities", :bool, true}, {"authentication_failure_close", :bool, true}, {"per_consumer_qos", :bool, true}, {"direct_reply_to", :bool, true}]}, {"cluster_name", :longstr, "rabbit@Boshans-MacBook-Pro"}, {"copyright", :longstr, "Copyright (C) 2007-2018 Pivotal Software, Inc."}, {"information", :longstr, "Licensed under the MPL.  See http://www.rabbitmq.com/"}, {"platform", :longstr, "Erlang/OTP 20.2.2"}, {"product", :longstr, "RabbitMQ"}, {"version", :longstr, "3.7.3"}], :none, false}

21:37:31.909 [error] gen_server <0.227.0> terminated with reason: {socket_error,timeout}
21:37:31.909 [error] CRASH REPORT Process <0.227.0> with 0 neighbours exited with reason: {socket_error,timeout} in gen_server:handle_common_reply/8 line 726
21:37:31.925 [error] gen_server <0.223.0> terminated with reason: {socket_error,timeout}
21:37:31.925 [error] Supervisor {<0.222.0>,amqp_connection_type_sup} had child main_reader started with amqp_main_reader:start_link(#Port<0.6798>, <0.223.0>, <0.225.0>, {method,rabbit_framing_amqp_0_9_1}, <<"client 127.0.0.1:50526 -> 127.0.0.1:5672">>) at <0.227.0> exit with reason {socket_error,timeout} in context child_terminated
21:37:31.925 [error] Supervisor {<0.222.0>,amqp_connection_type_sup} had child main_reader started with amqp_main_reader:start_link(#Port<0.6798>, <0.223.0>, <0.225.0>, {method,rabbit_framing_amqp_0_9_1}, <<"client 127.0.0.1:50526 -> 127.0.0.1:5672">>) at <0.227.0> exit with reason reached_max_restart_intensity in context shutdown
21:37:31.926 [error] CRASH REPORT Process <0.223.0> with 0 neighbours exited with reason: {socket_error,timeout} in gen_server:handle_common_reply/8 line 726
21:37:31.926 [error] Supervisor {<0.221.0>,amqp_connection_sup} had child connection started with amqp_gen_connection:start_link(<0.222.0>, {amqp_params_network,<<"guest">>,<<"guest">>,<<"/">>,"localhost",5672,0,0,0,infinity,none,[#Fun<amq..>,...],...}) at <0.223.0> exit with reason {socket_error,timeout} in context child_terminated
21:37:31.926 [error] Supervisor {<0.221.0>,amqp_connection_sup} had child connection started with amqp_gen_connection:start_link(<0.222.0>, {amqp_params_network,<<"guest">>,<<"guest">>,<<"/">>,"localhost",5672,0,0,0,infinity,none,[#Fun<amq..>,...],...}) at <0.223.0> exit with reason reached_max_restart_intensity in context shutdown
```

Using wireshark for debug, turns out after TCP connection established from amqp client to RabbitMQ server, 60s later client send tcp FIN to server, then the connection was closed.

Note that during those 60s, the connection is established successfully and I'm able to create a channel and publish message to an exchange.

Is this a bug in amqp client, or is there anything I did wrong?

Thanks in advance!

**Comment by pma** [comment]
@sunboshan I still using RabbitMQ 3.6.x and I've never seen this behaviour. But could be that 3.7 introduced a timeout on the server side for closing idle connections, which is worth exploring.

You can try setting a heartbeat on the connection, if this is the case, using:

```
iex(1)> AMQP.Connection.open(heartbeat: 30)
```

**Comment by sunboshan** [comment]
Thanks @pma. After opening connection with sending heartbeat every 30s, it's not closing the connection no more. After a quick search in Rabbit MQ Erlang client release note for 3.7.0, it could be [this PR](https://github.com/rabbitmq/rabbitmq-erlang-client/pull/86) that add the timeout. I'm new to Rabbit MQ, is this causing the connection close in 60s?

If that's the cause, should we always open the connection with a heartbeat less than 60s to avoid connection drop? Is that right way to use RabbitMQ?

**Comment by pma** [comment]
@sunboshan The default value of heartbeat was changed from 0 (disabled) to 10 in the Erlang client (see commit https://github.com/rabbitmq/rabbitmq-erlang-client/commit/ac77a4a59feaa3d7f6c78bb8ea49a990fc615f0a#diff-3d91045ee1a7022c8c8e1439bc8b10c4). Since the amqp Elixir lib is overriding the values (here: https://github.com/pma/wabbit/blob/master/lib/wabbit/connection.ex#L158) it should use the same defaults. Care to send a PR?

**Comment by ono** [comment]
Ah, good spot.

@sunboshan if you have time today you deserve to take the credit so please fix the value [here](https://github.com/pma/amqp/blob/master/lib/amqp/connection.ex#L87) and make PR. I will merge it, bump the version and release `1.0.1` immediately after that.

**Comment by ono** [comment]
Merged the PR and released 1.0.1. Thanks!

**Comment by damien-roche** [comment]
I'm confused. Does that mean this should not be happening now? I recently ran into the issue using amqp 1.0 and the above heartbeat option fixed it.
